File: 02.xtrem-t.d.d/src/main/java/money_problem/domain/Bank.java
package money_problem.domain;

import java.util.HashMap;
import java.util.Map;

public final class Bank {
    private final Map<String, Double> exchangeRates;

    private Bank(Map<String, Double> exchangeRates) {
        this.exchangeRates = exchangeRates;
    }

    public static Bank withExchangeRate(Currency from, Currency to, double rate) {
        var bank = new Bank(new HashMap<>());
        bank.addExchangeRate(from, to, rate);

        return bank;
    }

    public void addExchangeRate(Currency from, Currency to, double rate) {
        exchangeRates.put(keyFor(from, to), rate);
    }

    private static String keyFor(Currency from, Currency to) {
        return from + "->" + to;
    }

    public double convert(Money money, Currency to) throws MissingExchangeRateException {
        if (!canConvert(money.currency(), to)) {
            throw new MissingExchangeRateException(money.currency(), to);
        }
        return convertSafely(money, to);
    }

    // @TODO Return a Money here
    private double convertSafely(Money money, Currency to) {
        return money.currency() == to
                ? money.amount()
                : money.amount() * exchangeRates.get(keyFor(money.currency(), to));
    }

    private boolean canConvert(Currency from, Currency to) {
        return from == to || exchangeRates.containsKey(keyFor(from, to));
    }
}

File: 02.xtrem-t.d.d/src/main/java/money_problem/domain/functional/Bank.kt
package money_problem.domain.functional

import arrow.core.Either
import arrow.core.left
import arrow.core.right

typealias ExchangeRate = Double

class Bank private constructor(private val exchangeRates: Map<String, ExchangeRate>) {
    private fun convertSafely(money: Money, to: Currency): Money =
        if (money.currency == to) money
        else Money(money.amount * exchangeRates.getOrDefault(keyFor(money.currency, to), 0.0), to)

    private fun Currency.canConvert(to: Currency): Boolean =
        this == to || exchangeRates.containsKey(keyFor(this, to))

    fun addExchangeRate(from: Currency, to: Currency, rate: Double): Bank =
        Bank(
            exchangeRates.filterKeys { it != keyFor(from, to) } + (keyFor(from, to) to rate)
        )

    // Alternatives here: https://arrow-kt.io/learn/typed-errors/either-and-ior/
    fun convert(money: Money, currency: Currency): Either<MissingExchangeRateError, Money> = when {
        money.currency.canConvert(currency) -> convertSafely(money, currency).right()
        else -> MissingExchangeRateError(money.currency, currency).left()
    }

    companion object {
        fun withExchangeRate(from: Currency, to: Currency, rate: Double): Bank =
            Bank(HashMap()).addExchangeRate(from, to, rate)

        private fun keyFor(from: Currency, to: Currency): String = "$from->$to"
    }
}